cpp/src/arrow/compute/exec/expression_test.cc

```diff
@@ -1098,6 +1122,12 @@ TEST(Expression, CanonicalizeAnd) {
 
   // catches and_kleene even when it's a subexpression
   ExpectCanonicalizesTo(is_valid(and_(b, true_)), is_valid(and_(true_, b)));
+
+  auto ts = field_ref("ts_ns");
+  using namespace arrow_literals;
+  ExpectCanonicalizesTo(add(ts, literal(5min)), add(literal(5min), ts));
+  ExpectCanonicalizesTo(add(add(ts, literal(5min)), add(literal(5min), literal(5min))),
+                        add(add(add(literal(5min), literal(5min)), literal(5min)), ts));
```

Comment:
Wouldn't `add(literal(5min), literal(5min))` simplify to `add(literal(10min)` since all the args are literals?

Reply:
This is only checking the canonicalization pass, which (among other things) flattens and refolds chains with literals first. This sets up an easy situation for constant folding to do what you describe and eagerly evaluate constant calls. The distinct passes are an implementation detail of the main entry point `SimplifyWithGuarantee`, which might need to run cheap passes like constant folding multiple times if more expensive passes like guarantee substitution happen to yield a constant call.